### PR TITLE
Q# Package System

### LINKED ISSUE #883: Initial Q# Package System
# Q# Package System

## Definitions
* **package**: an exported or imported Q# project. 
* **package system**: the canonical and conventional way to export or import Q# projects, as packages.

Note the above constrained definitions -- we are not trying to build a registry or robust SemVer-based system at this point. 

## Objectives
* Unblock code-sharing for library authors.
* Provide alternative canonical Microsoft libraries that are not the standard library
* Provide a conventional system that supports the common use case of sharing publicly available code
* Don't immediately support private libraries, that can be a follow-on task
* Be able to construct a complete list of sources and fetch them before compilation begins

## User Stories
* As a hardware partner, I would like to publish an algorithm, as a Q# library, to showcase my hardware's capabilities.
* As a Q# author, I would like to import code without having to copy-paste it into my project.

## Potential Design
We initially only support shared code on GitHub. We use the GitHub REST API, which supports listing directories and cloning subfolders. This helps us avoid `libgit2`, which is impermissively licensed, and any native git implementations which may not work on all of our scenarios.

To import a Q# package, the user will add an entry to their manifest file's `dependencies` property. This will look similar to the `dependencies` property in a package.json. [We could even use their GitHub URL shorthand.](https://docs.npmjs.com/cli/v6/configuring-npm/package-json#github-urls), or we could just use GitHub URLs.

We will browse to that URL, look for a `qsharp.json` in the specified subfolder, or the root if no subfolder is specified, and parse the dependencies of that package. We will continue this process, browsing to them and parsing their manifests, with all dependencies until we have a complete graph of direct dependencies and transitive dependencies. At this stage, we will load all of these sources into memory and begin compilation as normal. In non-web scenarios, we may be able to cache these dependencies to disk, but in web scenarios, they may be cached in the state of the language service.

Related: #73 
```[tasklist]
### Tasks
- [ ] Tests: test dependency ordering edge cases with different language features enabled, etc
- [ ] Tests: test that entry points in libraries are not used
- [x] wire up all scenarios
- [x] work out TODOs on alex's branch
- [ ] telemetry
- [ ] Load transitive deps for qsc
- [ ] Load transitive deps for qsi
- [x] go-to-def and breaking _into_ things don't work
- [ ] https://github.com/microsoft/qsharp/pull/1651#discussion_r1646331561 change this test
```

### PR DESCRIPTION
**Pull request by sezna**
This PR:

1. Refactors the compiler API such that the caller passes in a constructed package store, allowing for dependencies to be pre-compiled and inserted before compiling user code
2. Introduces the `BuildableProgram` abstraction, which encapsulates the notion of "some user code that is ready to compile, along with a package store that has all of its dependencies compiled"
3. As a drive-by fix, fixes circuits for internal operations
4. Removes the notion of `Visibility` from the AST, since visibility is calculated via exports now. `internal` is still parsed without error for backwards compatibility, but is a no-op because `internal` is the default now.
5. Enforces visibility in the HIR, where items must be `Visibility::Public` to show up across packages as a `GlobalItem`
6. Adds export statements to the standard library to preserve the existing API with the new internal-by-default semantics
7. Inserts all packages into the user code's namespace tree with the prefix defined as the package alias in the `qsharp.json`

closes #883 

The Q# formatting CI stage will fail until #1692 goes in

This PR supersedes #1693

### CODE REVIEW

File: compiler/qsc_frontend/src/compile/tests.rs
Comment (swernli) [inline compiler/qsc_frontend/src/compile/tests.rs:968]: This doesn't seem right, but revising `@Unimplemented` relative to exports seems out of scope, especially since it is not used anywhere today (or even documented). Can you either comment out the test attribute (or skip) and then file a follow up issue on me?

File: language_service/src/state/tests.rs
Comment (swernli) [inline language_service/src/state/tests.rs:1851]: This seems odd, might be worth digging into why spans from dependencies are lost.

File: compiler/qsc_linter/src/linter/ast.rs
Comment (swernli) [inline compiler/qsc_linter/src/linter/ast.rs:64]: It occurs to me that one side effect of removing visibility from the AST is that we won't be able to add a lint warning in the future to tell people it's unnecessary and should be removed. Is the possibility of a future lint enough of a reason to keep it in?

File: pip/tests/test_project.py
Comment (swernli) [inline pip/tests/test_project.py:149]: Looks like a merge conflict got mistakenly checked in here?

File: samples/language/DemoProject/qsharp.json
Comment (swernli) [inline samples/language/DemoProject/qsharp.json:10]: do we want to check this in with a dependency to a personal repo? Would it make sense to have it in this repo, or does that get weird?

File: samples/language/DemoProject/src/Main.qs
Comment (swernli) [inline samples/language/DemoProject/src/Main.qs:13]: This sample triggers the "operation should be a function" lint, so probably worth making it `function` to avoid confusion.

File: wasm/src/project_system.rs
Comment (swernli) [inline wasm/src/project_system.rs:382]: It struck me how usefly this comment is to remind the reader that `BuildableProgram::new` actually does quite a bit of compiling... at the very least it's worth putting a description into the doc comment on `new` so folks can get that reminder on hover. As food for thought for future work: is there another name besides `new` that can convey how heavyweight this is without being too different from idomatic Rust?

File: wasm/src/language_service.rs
Comment (swernli) [inline wasm/src/language_service.rs:33]: This looks like it's leftover from an earlier iteration, it can be removed.

File: wasm/src/diagnostic.rs
Comment (swernli) [inline wasm/src/diagnostic.rs:56]: this change doesn't seem required and can be dropped.

File: vscode/test/suites/language-service/language-service.test.ts
Comment (swernli) [inline vscode/test/suites/language-service/language-service.test.ts:220]: If the removed `executeCommand("workbench.action.problems.focus")` wasn't needed here, is it still needed in the test case below (line 262)?

File: library/qs_source/core/core.qs
Comment (swernli) [inline library/qs_source/core/core.qs:61]: I think once your formatting change arrives this might need to be indented. Also, is it worth considering moving these to the top of the namespaces so when the file is opened fresh someone can see all the exports? That might get distracting for the longer lists, but would still be informative. Your call.

File: library/qs_source/src/std/core.qs
Comment (swernli) [inline library/qs_source/src/std/core.qs:109]: nitpick: newline before this rather than after.

File: library/qs_source/src/std/internal.qs
Comment (swernli) [inline library/qs_source/src/std/internal.qs:250]: nitpick: missing newline

File: library/qs_source/src/std/diagnostics.qs
Comment (swernli) [inline library/qs_source/src/std/diagnostics.qs:152]: nitpick: extra newline

File: library/qs_source/core/core.qs
Comment (swernli) [reply library/qs_source/core/core.qs:61]: Actually, never mind on moving these to the top. I realize now the exports are effectively for the purpose of making the callables "public" and later we'll have a std somewhere that re-exports things.

**Review by swernli — COMMENTED** [review]
Finished reviewing, changes look good, a few minor things called out. The only major thing is a possible side effect of removing visibility from the AST precluding future lints, but even that can be something we add back in later in a limited fashion if it makes sense to do so.

**Review by swernli — APPROVED** [review]

File: compiler/qsc/src/bin/qsc.rs
Comment (cesarzc) [inline compiler/qsc/src/bin/qsc.rs:112]: This block of logic grew considerably in this PR. It would be worth breaking it down in different functions so it can be more easily seen where `store`, `dependencies` and `source_map` come from.

File: compiler/qsc/src/bin/qsi.rs
Comment (cesarzc) [inline compiler/qsc/src/bin/qsi.rs:93]: Similarly to my other comment on `qsc`, it would be worth breaking down the logic here into multiple fucntions such that it's easier to see where the values of `store`, `dependencies` and `source_map` come from.

File: compiler/qsc/src/packages/tests.rs
Comment (cesarzc) [inline compiler/qsc/src/packages/tests.rs:128]: This test is called `missing_dependency`, but it doesn't seem to assert any errors so I think I'm not getting something. How can I see the missing dependency in this test?

File: compiler/qsc_project/src/project.rs
Comment (cesarzc) [inline compiler/qsc_project/src/project.rs:649]: Was this log macro intentionally left behind?

File: compiler/qsc_frontend/src/compile.rs
Comment (cesarzc) [inline compiler/qsc_frontend/src/compile.rs:186]: Are the stdlib and core considered legacy after these changes? What makes them legacy?

File: compiler/qsc_passes/src/capabilitiesck/tests_common.rs
Comment (cesarzc) [reply compiler/qsc_passes/src/capabilitiesck/tests_common.rs:53]: Looking at the code, std is used in these tests so I would be surprised if the tests passed without it.

File: katas/src/lib.rs
Comment (cesarzc) [inline katas/src/lib.rs:31]: After seeing this code repeated multiple times, I wonder if it would make sense to have utility functions like `qsc::PackageStore::with_core_lib` and `qsc::PackageStore::with_std_lib` to reduce the boilerplate code a bit.

**Review by cesarzc — APPROVED** [review]
I finished reviewing and the changes look good. I left a few comments with mostly non-blocking things, but it would be good to address them before merging.

**Pull request merged**
